### ISSUE TITLE
vimade and tmux split

### ISSUE THREAD

**Issue by lanox** [TaDaa/vimade#20]
Hi, 
After further troubleshooting I have narrowed it down to tmux.

So I saw in FAQ/Help that you mention tmux will not work if TERM is not set correctly.

**Problem Summery:**
In tmux split mode

One side I have vim and in another just terminal.
if I have multiple vim splits, vimade will dim the vim split that I am not in, however if I change to tmux split to terminal vimade will not dim the other vim pane.

inside tmux
echo $TERM is showing screen-256color

outside tmux
xterm-256color

**Comment by TaDaa** [comment]
Thanks for the info--

 I am seeing that in vim latest that ```termguicolors``` may not be working properly with tmux + ```screen-256color```.  If I remember correctly there also may be additional config required for tmux to work with termguicolors.

- Can you let me know the results of ```echo &termguicolors```.   Tmux may require additional configuration if you are using it.    You can tell if termguicolors is working properly by altering guibg highlights, -- e.g.  ```hi Normal guibg=#FFFFFF```  should change the background to white.  If it doesn't turn white, tmux requires additional config or you may need to disable termguicolors in your vimrc.

-  Another thing to try is setting ```export TERM=xterm-256color``` before running vim.  This should work even if termguicolors was set, but there will still likely be some color loss when fading.

**Comment by lanox** [comment]
No Problems, 

```echo &termguicolors``` returns ```1```
``` hi Normal guibg=#FFFFFF``` does change background to white.

I have tried setting ```export TERM=xterm-256color``` while in tmux and just before starting vi.

I should note as well that I am using neovim.

**Comment by TaDaa** [comment]
oh think I just realized that I completely misread the issue!

So vimade is working within a single split in tmux.  But you are expecting it to dim the last active/current window when you switch tmux splits?  I see this is an interesting idea/enhancement.  I think from the vimade standpoint there would need to be some way to detect focus changes since FocusGained and FocusLost don't seem to work consistently in terminals, but maybe there is another plugin that can be used?  I'll look into this some more later.

**Comment by TaDaa** [comment]
Any feedback / ideas welcome as well -- not an active tmux user, so if you have settings / plugins that are working to enable focus events let me know.

I think we will could probably expose a new command to fade/unfade the active window and instructions for binding it to FocusLost/FocusGained

**Comment by TaDaa** [comment]
Currently this should work in nvim now -- it requires a plugin to enable FocusGained and FocusLost events in tmux as well as setting the appropriate tmux config (```set -g focus-events on```) in your tmux config.

There are a few options for plugins that enable focus, but I have verified at the very least this one is working:
https://github.com/tmux-plugins/vim-tmux-focus-events


Additionally if tmux users can enable this functionality by adding the following to vimrc:
```
au FocusLost * call vimade#ForceFadel()
au FocusGained * call vimade#UnforceFade()
```

Let me know how it goes -- I will update the docs and readme with this later today

**Comment by TaDaa** [comment]
Changed vimade integration into this -- wasnt too happy with ForceFade naming..updated above as well
```
au! FocusLost * VimadeFadeActive
au! FocusGained * VimadeUnfadeActive
```

**Comment by lanox** [comment]
Hi @TaDaa 

That seem to be working however vimade seem to be very slow when loading vim.

this is what i got on startup time.

```
Slowest 10 plugins (out of 26)~
      vimade	306.796 
   [runtime]	278.758 
     gruvbox	20.323  
    coc.nvim	12.408  
    nerdtree	10.344  
     [vimrc]	8.170   
 vim-airline	4.921   
 delimitMate	3.646   
vim-devicons	2.739   
      vim-go	1.636   
```

**Comment by TaDaa** [comment]
Did this just start from latest change?  Also any buffers being preloaded during startup?  They can add to vimade startup time.

**Comment by lanox** [comment]
@TaDaa no it was quiet slow for me to start off, no I do not have any buffers preloaded, If you have any other ideas how to trouble shoot this I can help.

**Comment by TaDaa** [comment]
startup time should now be improved (at least according to profiler) in #29 -- feel free to reopen if needed.

**Comment by darianmorat** [comment]
@TaDaa Amazing plugin man! This is the best one by far.

But the plugin still needs to be improved in the startup time... It goes from 70-85ms to 350-460msms (approx 160 vimade + 160 python3). It's not that much really, but you can still notice it if u have a low startup time, I thought that a good idea would be not initializing the plugin in startup, but after it just when needed. For example, using lazyvim + telescope you can add this line: `cmd = "Telescope",` which will load telescope if I need to use it.

**Comment by TaDaa** [comment]
Thanks for flagging @darianmorat !  Adding lazyloading will be next :) (probably this weekend?)

For the increase you are seeing there seems to be a bug in Neovim (or maybe Vim).  Just calling `has('python')` or `has('python3')` causes the entire python file to load (that seems like overkill?).  Vimade unfortunately somehow also somehow also shows extra slowness (could also be a bug in startup time -- its odd the python3.vim file is basically the same value).

Commenting out these `has` calls reduced startup from 90ms to 1ms for me

**Comment by TaDaa** [comment]
That fix should prevent python3.vim from ever running (due to **vimade**) if you are in **neovim 0.8.0+ / lua-mode**

**Comment by darianmorat** [comment]
@TaDaa That was a killer change!
Is no longer calling python3.vim and vimade startup time is now less than 1ms... python3 was definitely the problem here, it's almost instant the startup time now, and is not even lazy loaded

**Comment by TaDaa** [comment]
ok great! Glad its better but that is crazy that python3 added so much latency:D.  Closing this bug out again, will still add lazy loading this weekend:)

**Comment by Fabrice-Bernes** [comment]
This is something I wanted to comment for a while now, but haven't had the time to do so. Let me know if I should open an issue instead of using this one.

Everything starting from commit [2dd0c37](https://github.com/TaDaa/vimade/commit/2dd0c37f75df6e009c9aa7356232b997820b1927) gives me the same behaviour that's described in the first comment :
* Switching between vim windows fades then properly (even if they show the same buffer, wich is awesome).
* Going from a vim pane to a any other tmux pane *DOES NOT* fade the active vim window like it should.
* Unfocusing the terminal emulator that's attached to tmux *DOES NOT* fade the active window either.
* Furthermore, running `:!au FocusLost VimadeFadeActive` doesn't fixes anything which is is weird.

It looks like the README doesn't reflects current behaviour for the `usecursorhold` option:

> [...] Setting this option disables the timer. [...]

The obvious fix is to replace line 92 (in the latest commit) in `autoload/vimade.vim` from:
```vimscript
    let g:vimade_features.has_timer_start = exists('*timer_start')
```
to:
```vimscript
    let g:vimade_features.has_timer_start = exists('*timer_start') && !g:vimade.usecursorhold
```
But I didn't make a pull request because you seem to have plans to add timer config options.
---
Unrelated, but thank you for this plugin, it's one of my must-haves.

**Comment by Fabrice-Bernes** [comment]
My proposal was dumb. This respects your naming convention better.
```diff
diff --git a/autoload/vimade.vim b/autoload/vimade.vim
index 8f3690c..b4288a1 100644
--- a/autoload/vimade.vim
+++ b/autoload/vimade.vim
@@ -602,7 +602,7 @@ function! vimade#FadePriority(priority)
 endfunction

 function! vimade#DeferredCheckWindows()
-  if g:vimade_features.has_timer_start
+  if g:vimade_features.has_timer_start && !g:vimade.usecursorhold
     " only disable the deferred on animation for lua (lua doesn't
     " only_these_windows)
     if exists('g:vimade_deferred_timer') || (exists('g:vimade_animation_running') && g:vimade_active_renderer == s:lua_renderer)
@@ -633,7 +633,7 @@ function! vimade#CheckWindows()
 endfunction

 function! vimade#StartAnimationTimer()
-  if g:vimade_features.has_timer_start
+  if g:vimade_features.has_timer_start && !g:vimode.usecursorhold
     let g:vimade_animation_running = 1
     if !exists('g:vimade_animation_timer')
       let g:vimade_animation_timer = timer_start(16, 'vimade#DoAnimations', {'repeat': -1})
```

**Comment by TaDaa** [comment]
Thanks for flagging @Fabrice-Bernes !  

I was able to reproduce the issue , the deferred logic I put in causes `vimade#FocusLost` pause to occur before the window is faded, I'll get this fixed this evening.

**Comment by Fabrice-Bernes** [comment]
Yes it is if fixed now. Thanks again

**Comment by TaDaa** [comment]
Great! Thank you for confirming
